starlite/asgi/routing_trie/traversal.py

```diff
@@ -30,37 +28,43 @@ def traverse_route_map(
     path_params: List[Tuple[str, Any]] = []
     path_components = [p for p in path.split("/") if p]
 
-    for i, component in enumerate(path_components):
+    for component in path_components:
         if component in current_node.child_keys:
             current_node = current_node.children[component]
             continue
 
-        if current_node.path_param_definition:
-            param_definition = current_node.path_param_definition
-            if param_definition.type is Path:
-                path_params.append((param_definition.name, normalize_path("/".join(path_components[i:]))))
-                return current_node, path_params, path
-
-            path_params.append((param_definition.name, param_definition.parser(component)))
-            current_node = current_node.children[PathParameterSentinel]
-
-        continue
+        if current_node.path_type_path_param_definition:
+            path_params.append(
+                (
+                    current_node.path_type_path_param_definition.name,
+                    normalize_path(path.removeprefix(current_node.path)),
+                )
+            )
+            return current_node, path_params, path
+
+        for param_definition in current_node.child_path_parameter_type_map.values():
+            try:
+                value = param_definition.parser(component)
+            except ValueError:
+                continue
+            path_params.append((param_definition.name, value))
+            current_node = current_node.children[param_definition]
+            break
```

Comment:
If there is going to be a performance impact in routing requests this is where it will come from, and will be relative to the number of adjacent path parameters that are children to the node.